src/typechecker/FStar.TypeChecker.Util.fs

```diff
@@ -93,61 +93,250 @@ let check_uvars r t =
+             (Warning_DeprecatedGeneric,
+              "Multiple decreases clauses on this definition; please remove the one on its declaration");
+          move_decreases d (pfx@sfx) (pfx'@sfx')
+        | Some (pfx, DECREASES d, sfx), None ->
+          move_decreases d (pfx@sfx) (U.comp_flags c')
+        | _ -> failwith "Impossible"
+  in
+  let extract_annot_from_body (lbtyp_opt:option<typ>)
+    : typ
+    * term
+    * bool
+    = let rec aux e
+        : typ * term * bool
+        = let e = SS.compress e in
+          match e.n with
+          | Tm_meta(e', m) ->
+            let t, e', recheck = aux e' in
+            t, { e with n = Tm_meta(e', m) }, recheck
+
+          | Tm_ascribed(e', (Inr c, tac_opt), lopt) ->
+            if U.is_total_comp c
+            then let t, lbtyp, recheck = reconcile_let_rec_ascription_and_body_type (U.comp_result c) lbtyp_opt in
+                 let e = { e with n = Tm_ascribed(e', (Inr (S.mk_Total t), tac_opt), lopt) } in
+                 lbtyp, e, recheck
+            else raise_error (Errors.Fatal_UnexpectedComputationTypeForLetRec,
+                              BU.format1 "Expected a 'let rec' to be annotated with a value type; got a computation type %s"
+                                                   (Print.comp_to_string c))
+                              rng
+
+          | Tm_ascribed(e', (Inl t, tac_opt), lopt) ->
+
+            let t, lbtyp, recheck = reconcile_let_rec_ascription_and_body_type t lbtyp_opt in
+            let e = { e with n = Tm_ascribed(e', (Inl t, tac_opt), lopt) } in
+            lbtyp, e, recheck
+
+          | Tm_abs _ ->
+            let bs, body, rcopt = U.abs_formals_maybe_unascribe_body false e in
+            let mk_comp t =
+              if Options.ml_ish()
+              then U.ml_comp t t.pos
+              else S.mk_Total t
+            in
+            let mk_arrow c = U.arrow bs c in
+            let rec aux body =
+              let body = SS.compress body in
+              match body.n with
+              | Tm_meta (body, m) ->
+                let t, body', recheck = aux body in
+                let body = { body with n = Tm_meta(body', m) } in
+                t, body, recheck
+
+              | Tm_ascribed (_, (Inl t, _), _) -> //no decreases clause here
+                begin
+                match lbtyp_opt with
+                | Some lbtyp ->
+                  lbtyp, body, false
+
+                | None ->
+                  let t = mk_arrow (mk_comp t) in
+                  t, body, true
+                end
+
+              | Tm_ascribed (body', (Inr c, tac_opt), lopt) ->
+                let tarr = mk_arrow c in
+                let tarr, lbtyp, recheck = reconcile_let_rec_ascription_and_body_type tarr lbtyp_opt in
+                let bs', c = un_arrow tarr in
+                if List.length bs' <> List.length bs
+                then failwith "Impossible"
+                else let subst = U.rename_binders bs' bs in
```

Comment:
This substitution seems redundant since `reconcile_let_rec_ascription_and_body_type` does not change the binders in `tarr`?

Reply:
Not really. bs' is a re-opening of bs. un_arrow does an opening and arrow closes it. so 
i.e., bs' <> bs in this sequence of operations

let bs, c = un_arrow t in
let t' = arrow bs c in
let bs', c' = un_arrow t' in
bs' <> bs